src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/PooledByteBufferWriter.cs

```diff
@@ -0,0 +1,150 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Buffers;
+using System.Diagnostics;
+
+namespace System.Net.ServerSentEvents
+{
+    internal sealed class PooledByteBufferWriter : IBufferWriter<byte>, IDisposable
```

Comment:
Could this use https://github.com/dotnet/runtime/blob/main/src/libraries/Common/src/System/Net/ArrayBuffer.cs?
(looks like the same idea at a quick glance)

Reply:
Looks similar, but AFAICT it doesn't implement IBufferWriter. FWIW this is duplicated from STJ so clearly we would benefit from a shared thing.

Reply:
Yep. The STJ type implements `PipeWriter` instead of `IBufferWriter` so we can't just share the source without taking an extra dependency in pre 9.0 targets.

Reply:
> Yep. The STJ type implements `PipeWriter` instead of `IBufferWriter` so we can't just share the source without taking an extra dependency in pre 9.0 targets.

And that's not a case of just having two partial files, one with this type implementing IBufferWriter and one with it also adding the PipeWriter derivation?

Reply:
What I had in mind re: `ArrayBuffer` is if we can replace this file with something along the lines of
```c#

internal sealed class PooledByteBufferWriter : IBufferWriter<byte>, IDisposable
{
	private ArrayBuffer _buffer = new(initialSize: 256, usePool: true);
	
	public void Advance(int count) => _buffer.Commit(count);

	public Memory<byte> GetMemory(int sizeHint = 0)
	{
		_buffer.EnsureAvailableSpace(sizeHint);
		return _buffer.AvailableMemory;
	}

	public Span<byte> GetSpan(int sizeHint = 0)
	{
		_buffer.EnsureAvailableSpace(sizeHint);
		return _buffer.AvailableSpan;
	}
	
	public ReadOnlyMemory<byte> WrittenMemory => _buffer.ActiveMemory;
	
	public int Capacity => _buffer.Capacity;
	
	public int WrittenCount => _buffer.ActiveLength;
	
	public void Reset() => _buffer.Discard(_buffer.ActiveLength);
	
	public void Dispose() => _buffer.Dispose();
}
```

Reply:
Thanks, this is a huge improvement.